src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,349 @@ pub mod policy;
+                    (Some(_), None) => Err(ValidationError::Other(format!(
+                        "unsatisfiable DNS name constraint: NC {} cannot match SAN {}",
+                        pattern.0, name.0
+                    ))),
+                    (None, _) => Err(ValidationError::Other(format!(
+                        "malformed DNS name constraint: {}",
+                        pattern.0
+                    ))),
+                }
+            }
+            (GeneralName::IPAddress(pattern), GeneralName::IPAddress(name)) => {
+                match (
+                    IPConstraint::from_bytes(pattern),
+                    IPAddress::from_bytes(name),
+                ) {
+                    (Some(pattern), Some(name)) => Ok(Applied(pattern.matches(&name))),
+                    (Some(_), None) => Err(ValidationError::Other(format!(
+                        "unsatisfiable IP name constraint: NC {:?} cannot match SAN {:?}",
+                        pattern, name,
+                    ))),
+                    (None, _) => Err(ValidationError::Other(format!(
+                        "malformed IP name constraints: {:?}",
+                        pattern
+                    ))),
+                }
+            }
+            _ => Ok(Skipped),
+        }
+    }
+
+    /// Apply the current name constraints (including those in the specified
+    /// extensions, if any) to the accumulated SAN set.
+    ///
+    /// On success (no constraint violations found), the new constraints
+    /// are additionally added to the name constraint set for future checks.
+    fn apply_and_accumulate(&mut self, extensions: &Extensions<'a>) -> Result<(), ValidationError> {
+        let new_constraints = match extensions.get_extension(&NAME_CONSTRAINTS_OID) {
+            Some(nc) => Some(nc.value::<NameConstraints<'a>>()?),
+            None => None,
+        };
+
+        for san in &self.sans {
+            // If there are no applicable constraints, the SAN is considered valid so the default is true.
+            let mut permit = true;
+            for nc in self.name_constraints.iter().chain(new_constraints.iter()) {
+                if let Some(permitted_subtrees) = &nc.permitted_subtrees {
+                    for p in permitted_subtrees.unwrap_read().clone() {
+                        let status = self.apply_inner(&p.base, san)?;
+                        if status.is_applied() {
+                            permit = status.is_match();
+                            if permit {
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                if !permit {
+                    return Err(ValidationError::Other(
+                        "no permitted name constraints matched SAN".into(),
+                    ));
+                }
+            }
+
+            for nc in self.name_constraints.iter().chain(new_constraints.iter()) {
```

Comment:
Should this just be a singe loop, isn't of 2x loops?

Reply:
Yep, flattening.